### DISCUSSION TITLE
Is FileCache async?

### DISCUSSION THREAD

**Discussion by ludaavics** [obendidi/httpx-cache#45] [Q&A]
Hi,

Reading through [this](https://github.com/obendidi/httpx-cache/blob/main/httpx_cache/cache/file.py) am I correct in understanding that the use of async lock makes getting a large number of already-cached responses essentially sequential?

Thanks!

**Comment by obendidi** [comment]
I think you're right :thinking: 

the async_lock indeed blocks the access to the fs even if it's reading different files, the idea behind it was to avoid a race condition on r/w the same file.

Thanks for the catch

**Reply by ludaavics** [reply]
Right... if i'm thinking about this correctly, that implies large performance degradation compared to no caching? Ideally, we would want to use a file specific async lock?

Asking my friend Prof. Google,  [filelock](https://py-filelock.readthedocs.io/en/latest/index.html) acquires a lock on a specific file, but [doesn't support asyncio](https://github.com/tox-dev/py-filelock/issues/78). However:
 - i think it can't be worse than current implementation? though might not be a whole lot better
 - [these guys](https://github.com/nipype/pydra/pull/460/files#diff-61bb2ea1e92f5103191ee9b1ee2bc8395023cbb15cb15e903ae0d946bd077e55R896) seem to have  an asyncio workaround. Prob not as clean a bottoms up implementation, but prob significantly better than current implementation?

**Comment by ludaavics** [comment]
Thinking about this more, @obendidi, why do we care about race condition on the same file?
Presumably the two "racers" will have received the same http response?

**Reply by obendidi** [reply]
In the unlikely edge case where we send the same request twice, and if the response has a large payload, it is very possible that the second request will try to read the file while the first request has not finished writing it

**Reply by ludaavics** [reply]
In that case, instead of the file lock, what do you think of catching any read exception and handling it by re-issuing the http request ?

**Reply by ludaavics** [reply]
I guess in your example of large payload, the second request could attempt to write before the first request is done, so the write lock should stay. So writing could still end up close to sequential (in the case of a large number of quick http requests) but at least reading from cache would be fully asynchronous.

**Reply by obendidi** [reply]
yep, and it seems there is a good library for that: https://github.com/aio-libs/aiorwlock

>...., The read lock may be held simultaneously by multiple reader tasks, so long as there are no writers. The write lock is exclusive.

**Reply by obendidi** [reply]
made a PR with a possible fix, feel free to have a look
https://github.com/obendidi/httpx-cache/pull/48

**Comment by ludaavics** [comment]
In that case, instead of the file lock, what do you think of catching any
read exception and handling it by re-issuing the http request ?
